lib/web_of_science/process_pubmed.rb

```diff
@@ -0,0 +1,56 @@
+module WebOfScience
+
+  # This complements the WebOfScience::Harvester
+  module ProcessPubmed
+
+    # For WOS-records with a PMID, try to enhance them with PubMed data
+    def pubmed_additions(records)
+      records.select { |record| record.pmid.present? }.each do |record|
+        begin
+          pub = Publication.find_by(wos_uid: record.uid)
+          pub.pmid = record.pmid
+          pub.save
+          pubmed_addition(record, pub) if record.database != 'MEDLINE'
+        rescue StandardError => err
+          message = "pubmed_additions failed for #{record.uid}"
+          NotificationManager.error(err, message, self)
+        end
+      end
+    end
+
+    # For WOS-record that has a PMID, fetch data from PubMed and enhance the pub.pub_hash with PubMed data
+    # @param [WebOfScience::Record] record
+    # @param [Publication] pub
+    # @return [void]
+    def pubmed_addition(record, pub)
+      pubmed_record = PubmedSourceRecord.for_pmid(record.pmid)
+      if pubmed_record.nil?
+        pubmed_missing(record, pub)
+      else
+        pub.pub_hash.reverse_update(pubmed_record.source_as_hash)
+        pub.save
+      end
+    rescue StandardError => err
+      message = "#{record.uid}, PubmedSourceRecord failed, PMID: #{record.pmid}"
+      NotificationManager.error(err, message, self)
+    end
+
+    # For WOS-record that has a PMID, cleanup our data when it does not exist on PubMed
+    # @param [WebOfScience::Record] record
+    # @param [Publication] pub
+    # @return [void]
+    def pubmed_missing(record, pub)
+      WebOfScience.logger.warn "#{record.uid}, PubmedSourceRecord missing, PMID: #{record.pmid}"
+      return unless PubmedClient.working?
```

Comment:
this makes another round trip to Pubmed each time, which seems uncessary and may start to get expensive particularly when it actually being down has proved to be extremely rare... 

is there any way to tell from the PubmedSourceRecord.for_pmid call before this if the nil result was because nothing was found or because Pubmed was down?

Reply:
Yes, `working?` makes a round-trip.  But if it is only rarely expensive, that doesn't seem like a dealbreaker.

Reply:
yeah, i realize that'll probably be rare too (hopefully, assuming WoS data is mostly correct) so not a dealbreaker

Reply:
In the partial WOS-harvest, there were hundreds of bogus PMID.  It would be poor of us to mirror bad data in our Pub and Pub-Ids.  The round trip cost to check on the status of PubMed API seems to be essential though to prevent removing good PMID data.  There are a multitude of cost/benefit aspects to this, but mostly data integrity seems more important than performance hits.

Reply:
If possible, can we get a logging (ideally at aggregate level) of how many bad PMIDs we detect?  This seems important to characterize our confidence/skepticism about WOS PMIDS and might be useful feedback to Clarivate.